### PR TITLE
Backend with WIP Frontend

### LINKED ISSUE #6: Initial User Management
Django App: `core`

### Backend Tasks
* [x] #11
   * E-Mail (mandatory)
   * Markdown bio (optional)
   * Institution (optional)
   * GitHub Handle (optional)
   * Twitter Handle (optional)
* [x] #16
* [x] #18
* [x] #17
* [x] #24
* [x] #37
* [x] Override the default login form of Django

### Frontend Tasks
* [x] Landing page (not logged in) w/ button to login & button to register #31
* [x] Landing page (logged in) w/ button to logout #31
* [x] User Registration Modal #31

### LINKED ISSUE #11: Extend the default Django user with additional attributes for
new attributes should be:
 - E-Mail (mandatory)
 - Markdown bio (optional)
 - Institution (optional)
 - ~~GitHub Handle (optional)~~ moved to #8 
-  ~~Twitter Handle (optional)~~ moved to #8

### LINKED ISSUE #16: Create model and extend the admin view for the Users Model

### LINKED ISSUE #18: Create form for serving the "Register new user"-form to the frontend and processing submitted forms

### PR DESCRIPTION
**Pull request by wurstbroteater**
#### Backend
 - resolves #11
 - resolves #16
 - resolves #18

# !!THIS SHOULD BE MERGED BEFORE file-upload BRANCH !!
This PR includes outdated files from file-upload branch. You have to upate user-management branch with file-upload branch or merge  user-management before file-upload into main

## Migrations
 Can be done via `makeMigrate.sh` or `.bat`)
Sometimes this will throw errors after fresh pull. If so try to delete all migrations folders and the db.sqlite3 file and retry it

## Start
 - You can start the backend (this includes migrations)
 - Create superuser (E.g email= "a@b.de", password= "123")
 - Login in the admin panel on `http://localhost:8000/admin` 
 - Here you can create, alter, remove users
 - Click on a user and verify the user attributes
 - Change your password 
 - You can verify that every page except the admin page is rest based and sends json
 - Logout 
 - You can retrieve JWT access and refresh tokens with sending a POST request to  `http://localhost:8000/auth/login/`  with the following JSON:
```json
{"email": "a@b.de", "password" : "123"}
```
 - If a the access token is expried, you can retrieve a new one with sending the refresh token as POST request to  `http://localhost:8000/auth/refresh/` 
 -  You can view all users (also as JSON) in `http://localhost:8000/users/`
 - You can register a new user with sending a POST request to `http://localhost:8000/auth/register/` with the following JSON:
  ```json
{
    "email": "new@us.er",
    "password": "42a",
    "is_active": true,
    "is_staff": false
}
```

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (h3ssto) [inline .vscode/settings.json]: Please remove `.vscode` and all related files from the repository by adding it to the `.gitignore`-file and using `git rm --cached -r .vscode`.

File: backend/.gitignore [hunk 2]
```diff
@@ -158,3 +158,6 @@ dmypy.json
```
Comment (h3ssto) [inline backend/.gitignore]: I would suggest moving the `.gitignore` to the root directory of the project and unifying it.

File: backend/core/fileupload/migrations/0001_initial.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+# Generated by Django 3.2.8 on 2021-11-15 08:44
```
Comment (h3ssto) [inline backend/core/fileupload/migrations/0001_initial.py]: Please remove the `migrations/*` files from the repository as they only make sense for maintaining an existing db.

File: backend/core/fileupload/migrations/0002_auto_20211115_0914.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# Generated by Django 3.2.8 on 2021-11-15 09:14
```
Comment (h3ssto) [inline backend/core/fileupload/migrations/0002_auto_20211115_0914.py]: as above

File: backend/core/fileupload/migrations/0003_rename_user_file_owner.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 3.2.8 on 2021-11-15 09:22
```
Comment (h3ssto) [inline backend/core/fileupload/migrations/0003_rename_user_file_owner.py]: as above

File: backend/core/fileupload/migrations/0004_alter_file_owner.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# Generated by Django 3.2.8 on 2021-11-15 09:24
```
Comment (h3ssto) [inline backend/core/fileupload/migrations/0004_alter_file_owner.py]: as above

File: backend/core/user/migrations/0001_initial.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Generated by Django 3.2.8 on 2021-11-15 08:45
```
Comment (h3ssto) [inline backend/core/user/migrations/0001_initial.py]: as above

File: backend/makeMigrate.sh [hunk 8]
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
```
Comment (h3ssto) [inline backend/makeMigrate.sh:1]: I would not use this shebang but rather `#!/usr/bin/env python3`.

File: backend/yarn.lock [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (h3ssto) [inline backend/yarn.lock:1]: I suspect that this files also does not belong in the repository.

File: frontend/yarn.lock [hunk 10]
```diff
@@ -1098,13 +1098,20 @@
   dependencies:
```
Comment (h3ssto) [inline frontend/yarn.lock:1098]: Same as above, this file should be removed from the repository.

**Review by h3ssto — CHANGES_REQUESTED** [review]
Nice work in general.
Please move `.gitignore` to the root directory and add proper entries for Python, Django, Node, yarn, e.g.: https://www.toptal.com/developers/gitignore/api/python,django,node,yarn,vs,pycharm+all

This should also ignore migrations and `yarn.lock` files in the future. After amending the `.gitignore` file I suggest executing
```bash
git rm -r --cached .
git add .
```
to enforce the new `.gitignore` rules.

Minor comments:
* We should unify capitalization of files for frontend and backend respectively, I would suggest
   * lowerCamelCase for files and typescript methods
   * UpperCamelCase for classes
   * snake_case for Python methods
   * prefixing tests with `test_`

File: backend/yarn.lock (hunk 9 above)
Comment (wurstbroteater) [reply backend/yarn.lock:1]: `yarn.lock` should be kept in SC, see https://github.com/yarnpkg/yarn/issues/1583#issuecomment-257544217

File: backend/makeMigrate.sh (hunk 8 above)
Comment (wurstbroteater) [reply backend/makeMigrate.sh:1]: this does not work on my ubuntu/Pycharm config

File: frontend/yarn.lock (hunk 10 above)
Comment (h3ssto) [reply frontend/yarn.lock:1098]: `yarn.lock` is not a temporary lock-file and should therefore be kept in VCS; https://classic.yarnpkg.com/en/docs/migrating-from-npm

**Review by h3ssto — APPROVED** [review]

**Pull request merged**